plugin/pkg/rqdata/rqdata.go

```diff
@@ -0,0 +1,220 @@
+package rqdata
+
+import (
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coredns/coredns/plugin/pkg/dnstest"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+type requestFunc func(state request.Request) string
+
+// Mapping define the mapping between 'name' of data and the way to extract that data from the Request
+// it also defines what will be the empty value returned if the data behind the name is empty.
+// it is pretty static, and you should need to instantiate only once
+type Mapping struct {
+	replacements map[string]requestFunc
+	emptyValue   string
+}
+
+// Extractor implements a Value(name) (value, valid) function
+// which allow to extract data from an existing DNS Request(or state)
+type Extractor struct {
+	state     request.Request
+	requester *Mapping
+}
+
+//NewExtractor return a new Extractor based on the Mapping and the Request provided
```

Comment:
no space between // and comment. Also no closing dot (please check the rest of the code)